internal/kafka/command_partition.go

```diff
@@ -6,35 +6,43 @@ import (
 
 	"github.com/spf13/cobra"
 
-	"github.com/confluentinc/kafka-rest-sdk-go/kafkarestv3"
-
 	pcmd "github.com/confluentinc/cli/v3/pkg/cmd"
+	"github.com/confluentinc/cli/v3/pkg/config"
 )
 
 type partitionCommand struct {
 	*pcmd.AuthenticatedCLICommand
 }
 
-func newPartitionCommand(prerunner pcmd.PreRunner) *cobra.Command {
+func newPartitionCommand(cfg *config.Config, prerunner pcmd.PreRunner) *cobra.Command {
 	cmd := &cobra.Command{
 		Use:         "partition",
 		Short:       "Manage Kafka partitions.",
-		Annotations: map[string]string{pcmd.RunRequirement: pcmd.RequireOnPremLogin},
+		Annotations: map[string]string{pcmd.RunRequirement: pcmd.RequireNonAPIKeyCloudLoginOrOnPremLogin},
 	}
 
-	c := &partitionCommand{pcmd.NewAuthenticatedWithMDSCLICommand(cmd, prerunner)}
-	c.PersistentPreRunE = prerunner.InitializeOnPremKafkaRest(c.AuthenticatedCLICommand)
+	c := &partitionCommand{}
+
+	if cfg.IsCloudLogin() {
+		c.AuthenticatedCLICommand = pcmd.NewAuthenticatedCLICommand(cmd, prerunner)
 
-	cmd.AddCommand(c.newDescribeCommand())
-	cmd.AddCommand(c.newReassignmentCommand())
-	cmd.AddCommand(c.newListCommand())
+		cmd.AddCommand(c.newDescribeCommand())
+		cmd.AddCommand(c.newListCommand())
+	} else {
+		c.AuthenticatedCLICommand = pcmd.NewAuthenticatedWithMDSCLICommand(cmd, prerunner)
+		c.PersistentPreRunE = prerunner.InitializeOnPremKafkaRest(c.AuthenticatedCLICommand)
+
+		cmd.AddCommand(c.newDescribeCommandOnPrem())
+		cmd.AddCommand(c.newListCommandOnPrem())
+		cmd.AddCommand(c.newReassignmentCommand())
+	}
 
 	return cmd
 }
 
-func parseLeaderId(leader kafkarestv3.Relationship) int32 {
-	index := strings.LastIndex(leader.Related, "/")
-	idStr := leader.Related[index+1:]
+func parseLeaderId(related string) int32 {
+	index := strings.LastIndex(related, "/")
+	idStr := related[index+1:]
```

Comment:
Might be more readable to split by "/" and take the last element in the returned array.